src/Functions/fromReadableSize.cpp

```diff
@@ -0,0 +1,265 @@
+#include <limits>
+#include <boost/algorithm/string/case_conv.hpp>
+#include <boost/convert.hpp>
+#include <boost/convert/strtol.hpp>
+
+#include <Columns/ColumnsNumber.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/IFunction.h>
+#include "base/types.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int TOO_FEW_ARGUMENTS_FOR_FUNCTION;
+    extern const int TOO_MANY_ARGUMENTS_FOR_FUNCTION;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int BAD_ARGUMENTS;
+}
+
+namespace
+{
+
+    const std::unordered_map<std::string_view, UInt64> size_unit_to_bytes =
+    {
+        {"b", 1},
+        // ISO/IEC 80000-13 binary units
+        {"kib", 1024},                  // 1024
+        {"mib", 1048576},               // 1024 * 1024
+        {"gib", 1073741824},            // 1024 * 1024 * 1024
+        {"tib", 1099511627776},         // 1024 * 1024 * 1024 * 1024
+        {"pib", 1125899906842624},      // 1024 * 1024 * 1024 * 1024 * 1024
+        {"eib", 1152921504606846976},   // 1024 * 1024 * 1024 * 1024 * 1024 * 1024
+
+        // SI units
```

Comment:
Hallo, thanks for the detailed review & the many pointers, really appreciate it!

Indeed, I understood from the examples in #61548 that we wanted to handle the different units with a single function, I've now split them in matching `from*Size` functions.

> Side node 1: One could argue that fromReadableSize should only be able to parse text produced by formatReadableSize, i.e. text with a single whitespace between the number and the unit, case-insensitive parsing) but I think that would be overkill and not very human-friendly.

Agree, I've made the parsing insensitive to both casing & the number of spaces between the size number & unit. What I'm strictly forbidding is leading & trailing whitespace as discussed in other comments, if you think we should also relax those constraints please let me know.

> Side note 2: By convention, many parsing ClickHouse functions provide also OrNull and OrThrow variants, see e.g. parseDateTime ([docs](https://clickhouse.com/docs/en/sql-reference/functions/type-conversion-functions#parsedatetimeorzero)). If you like to add these for fromReadableSize I would be very happy. If not, it is also okay.

Absolutely, I'll add them now. The default behaviour in this case is to throw, checking the flavours of `parseDateTime` I see we have `orNull` and `orZero`, as we're returning a string here It might make sense to implement `orNull` and `orEmpty` (returning an empty string). Does that make sense for you?

Thanks!

Reply:
After further inspection of `formatReadableSize` I've realized that it can accept & render decimal & negative bytes so I've also changed the output type from `UInt64` to `Float64` to accommodate this.

For very large values (100s of PBs+) we're starting to see some precision loss and TBH I'm not sure what'd be the best way to accommodate this larger values completely accurately. If this is an issue for this initial implementation any guidance would be much appreciated. Thanks!

Reply:
> Absolutely, I'll add them now. The default behaviour in this case is to throw, checking the flavours of parseDateTime I see we have orNull and orZero so I'll do the same here, if that makes sense to you

Yes, that makes sense!

> After further inspection of formatReadableSize I've realized that it can accept & render decimal & negative number of bytes so I've also changed the output type from UInt64 to Float64 to accommodate this.

Negative outputs sound silly - 1. it is hard to think of a real-word equivalent (compare statement "this object weights negative 3 kilogram" - doesn't make sense, at least in this universe) and 2. `formatReadable*` never return negative values. Proposal: Treat a negative value as error \+ make the function return `UInt64` instead of `Float64` again.

Reply:
> Negative outputs sound silly - 1. it is hard to think of a real-word equivalent (compare statement "this object weights negative 3 kilogram" - doesn't make sense, at least in this universe) and 2. `formatReadable*` never return negative values. Proposal: Treat a negative value as error + make the function return `UInt64` instead of `Float64` again.

Got it, thanks! Added check that will raise an error on negative numbers and changed the output type back to `UInt64`, rounding decimals up.